### PR TITLE
Ability to switch usb debug mode to wireless

### PR DESCRIPTION
**Pull request by STFBEE**
Useful feature when you need to pick up phone from table and do something with it without usb cable while debugging

### CODE REVIEW

File: adbe/version.txt [hunk 1]
```diff
@@ -1 +1 @@
-2.5.4
\ No newline at end of file
+2.5.5
```
Comment (ashishb) [inline adbe/version.txt:1]: Please revert the change to `version.txt`. This file contains the version which is only incremented at the time of the release.

File: adbe/adb_enhanced.py [hunk 2]
```diff
@@ -1531,3 +1531,19 @@ def _error_if_min_version_less_than(min_acceptable_version, device_serial=None):
...
     return qemu is not None and qemu.strip() == '1'
+
+
+def reconnect_wireless():
+    code, result, stderr = execute_adb_shell_command2("ip address")
+    if code != 0:
+        print_error('Failed to switch device to wireless debug mode, stderr: %s' % stderr)
+        return
+
+    ip = re.findall(r"inet ([\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}).*wlan0$", result, re.MULTILINE)[0]
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1542]: `re.findall(r"inet ([\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}).*wlan0$", result, re.MULTILINE)` is not guaranteed to return a result.
For example, when I tried without connecting the device, it failed with

```
Traceback (most recent call last):
  File "./adbe/main.py", line 390, in <module>
    main()
  File "./adbe/main.py", line 335, in main
    adb_enhanced.reconnect_wireless()
  File "/private/tmp/adb-enhanced/adbe/adb_enhanced.py", line 1542, in reconnect_wireless
    ip = re.findall(r"inet ([\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}).*wlan0$", result, re.MULTILINE)[0]
IndexError: list index out of range
````

File: docs/README.rst [hunk 3]
```diff
@@ -65,6 +65,10 @@ Device configuration
 
 -  Take a video ``adbe screenrecord video.mp4 # Press ^C when finished``
 
+-  Switch Debug to Wireless mode
+
+``adbe wireless``
```
Comment (ashishb) [inline docs/README.rst:70]: Thanks. I just tried and it works seamlessly with a single device connected. Can you think of a better sub-command name? For example, `adbe enable wireless debugging`.

In the future, we can `adbe disable wireless debugging` as well.

**Comment by ashishb** [comment]
Thanks for your contribution. There is no issue tracking this. Feel free to open one and link it here.

**Review by ashishb — COMMENTED** [review]
Two minor issues and then this is ready for merge.

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (STFBEE) [reply adbe/adb_enhanced.py:1542]: Can you provide results of `adb shell ip address` from your phone (lines, that contains wlan interface info)?

File: docs/README.rst (hunk 3 above)
Comment (STFBEE) [reply docs/README.rst:70]: It's too long, you don't think? I use shell script with same functionality with short alias `adbw`, so it's very fast and simple to move device to wireless debug mode

Comment (STFBEE) [reply docs/README.rst]: Command added

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (ashishb) [reply adbe/adb_enhanced.py:1544]:
```
$ adb shell ip address
error: device offline
```

File: docs/README.rst (hunk 3 above)
Comment (ashishb) [reply docs/README.rst]: You are right that this is long and I am open to alternatives but `adbe wireless` is too short. It can even mean that you are turning the wireless (Wi-Fi?) on/off of the device. And not the ADB debugging connection.

Comment (STFBEE) [reply docs/README.rst]: First of all, user always can set his own aliases in bash/cmd environment. So we can set more informative commands and add shortcuts later

Comment (ashishb) [reply docs/README.rst]: Perfect. We are on the same page.

File: adbe/adb_enhanced.py (hunk 2 above)
Comment (ashishb) [reply adbe/adb_enhanced.py:1544]: Did you get a chance to fix this?

Comment (STFBEE) [reply adbe/adb_enhanced.py:1544]: oh, haven't seen an notification about your comment. I know how to fix this from terminal: `adb disconnect`, but i have no idea what to do with it if it appears at runtime

Comment (ashishb) [reply adbe/adb_enhanced.py:1544]: How about failing gracefully by doing the same thing you are doing a few statements above?

```
print_error('Failed to switch device to wireless debug mode, stderr: %s' % stderr)
```

Comment (STFBEE) [reply adbe/adb_enhanced.py:1544]: hm, no idea what to do: without devices i've got
'Failed to switch device to wireless debug mode, stderr: adb.exe: no devices/emulators found'
if device is offline, i've got
'Failed to switch device to wireless debug mode, stderr: adb.exe: device offline'

Comment (STFBEE) [reply adbe/adb_enhanced.py:1544]: Can't reproduce your case

Comment (STFBEE) [reply adbe/adb_enhanced.py:1549]: ok, finally i found it (by chance, lol). fixed

Comment (ashishb) [reply adbe/adb_enhanced.py:1549]: Thanks.

File: adbe/adb_enhanced.py [hunk 4]
```diff
@@ -1531,3 +1533,43 @@ def _error_if_min_version_less_than(min_acceptable_version, device_serial=None):
...
+
+def enable_wireless_debug():
+    code, result, stderr = execute_adb_shell_command2("ip address")
+    if code != 0:
+        print_error('Failed to switch device to wireless debug mode, stderr: %s' % stderr)
+        return
+
+    matching = re.match(r"inet ([\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}\.[\d]{1,3}).*wlan0$", result)
+    if matching is None:
+        print_error('Failed to switch device to wireless debug mode')
```
Comment (ashishb) [inline adbe/adb_enhanced.py:1546]: Probably `print_error_and_exit`?

**Review by ashishb — APPROVED** [review]

**Comment by ashishb** [comment]
> trying to fix more autocheck issues

If it is only cosmetic, I can fix it later as well.

**Comment by ashishb** [comment]
I tested it after merging and I am unable to make it work either for the emulator or the real device. Are you sure, `enable wireless debugging` is setup correctly? In my case, it is failing to find the IPv4 IP address via your regex to enable wireless debugging here - https://github.com/ashishb/adb-enhanced/blob/3888aaa590149eeda6cdd0d034e0c8ab99b8edfb/adbe/adb_enhanced.py#L1570

@STFBEE does it work for you?

**Comment by STFBEE** [comment]
Hmmm, I don't remember what that line was for, like to make sure the phone was connected to wifi and not in air mode or cellular network. I'll create a new MR with fix in the near future
